src/MonitorProfiler/EventProvider/ProfilerEvent.h

```diff
@@ -0,0 +1,179 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#pragma once
+
+#include "cor.h"
+#include "corprof.h"
+#include "com.h"
+#include "EventTypeMapping.h"
+#include <vector>
+#include <string>
+#include <memory>
+
+/// <summary>
+/// Helper class used to write EventSource data.
+/// Initialize is used to declare the data types and fill the COR_PRF_EVENTPIPE_PARAM_DESC structure.
+/// WritePayload is used to create COR_PRF_EVENT_DATA and write the actual data.
+/// We rely on variadic templates and template specialization for both of the above.
+///
+/// IMPORTANT All data to WritePayload must be valid until the data is written by ICorProfiler12, not just assigned to COR_PRF_EVENT_DATA.
+/// The profiler API uses memory addresses even for primitive types.
+/// </summary>
+template<typename... Args>
+class ProfilerEvent
+{
+    friend class ProfilerEventProvider;
+public:
+    HRESULT WritePayload(const Args&... args);
+
+private:
+    ProfilerEvent(ICorProfilerInfo12* profilerInfo);
+
+    template<size_t index, typename T, typename... TArgs>
+    HRESULT Initialize(const WCHAR* (&names)[sizeof...(Args)]);
+
+    template<size_t index>
+    HRESULT Initialize(const WCHAR* (&names)[sizeof...(Args)]);
+
+    template<size_t index, typename T, typename... TArgs>
+    HRESULT WritePayload(COR_PRF_EVENT_DATA* data, const T& first, TArgs... rest);
+
+    template<size_t index, typename T = tstring, typename... TArgs>
```

Comment:
I know we talked about why this is written this way yesterday, but would you want to leave a comment since it's not obvious to anyone who reads it later?